File: backend/src/main/java/com/ecommerce/app/catalog/application/mapper/ProductoMapper.java
package com.ecommerce.app.catalog.application.mapper;

import com.ecommerce.app.catalog.application.dto.request.ActualizarProductoRequest;
import com.ecommerce.app.catalog.application.dto.request.CrearProductoRequest;
import com.ecommerce.app.catalog.application.dto.response.ProductoBasicoResponse;
import com.ecommerce.app.catalog.application.dto.response.ProductoResponse;
import com.ecommerce.app.catalog.domain.model.Categoria;
import com.ecommerce.app.catalog.domain.model.Producto;
import org.springframework.stereotype.Component;

@Component
public class ProductoMapper {

    public ProductoBasicoResponse toBasicoResponse(Producto producto) {
        if (producto == null) {
            return null;
        }

        return new ProductoBasicoResponse(
                producto.getId(),
                producto.getNombre(),
                producto.getSku(),
                producto.getPrecio(),
                producto.getStock(),
                producto.getImagenNombre()
        );
    }
    
    public ProductoResponse toResponse(Producto producto) {
    	if (producto == null) {
    		return null;
    	}
    	
    	return new ProductoResponse(
    			producto.getId(),
    			producto.getNombre(),
    			producto.getSku(),
    			producto.getDescripcion(),
    			producto.getPrecio(),
    			producto.getStock(),
    			producto.getImagenNombre(),
    			producto.isActivo(),
    			producto.getCategoria().getId(),
    			producto.getCategoria().getNombre()
    	);
    }
    
    public Producto toEntity(CrearProductoRequest request, Categoria categoria) {
    	if (request == null) {
    		return null;
    	}
    	
    	Producto producto = new Producto();
    	producto.setNombre(request.nombre());
    	producto.setSku(request.sku());
    	producto.setDescripcion(request.descripcion());
    	producto.setPrecio(request.precio());
    	producto.setStock(0);
    	producto.setImagenNombre(request.imagenNombre());
    	producto.setActivo(true);
    	producto.setCategoria(categoria);
    	
    	return producto;    	
    }

	public void updateEntity(Producto producto, ActualizarProductoRequest request, Categoria categoria) {
		if (producto == null || request == null) {
			return;
		}

		producto.setNombre(request.nombre());
		producto.setSku(request.sku());
		producto.setDescripcion(request.descripcion());
		producto.setPrecio(request.precio());
		producto.setImagenNombre(request.imagenNombre());
		producto.setActivo(request.activo());
		producto.setCategoria(categoria);
	}
}

File: frontend/src/app/features/catalog/infrastructure/mappers/producto.mapper.ts
import { Producto } from '../../domain/models/producto.model';
import { ProductoResponseDto, CrearProductoRequestDto, ActualizarProductoRequestDto } from '../dtos/producto.dto';
import { RegistrarProductoData, EditarProductoData } from '../../domain/repositories/producto.repository';

export class ProductoMapper {
  public static toDomain(dto: ProductoResponseDto): Producto {
    return new Producto(
      dto.id,
      dto.nombre,
      dto.sku,
      dto.descripcion,
      dto.precio,
      dto.stock,
      dto.imagenNombre || null,
      dto.activo,
      dto.categoriaId,
      dto.categoria
    );
  }

  public static toDomainList(dtos: ProductoResponseDto[]): Producto[] {
    return dtos.map((dto) => this.toDomain(dto));
  }

  public static toCreateRequest(data: RegistrarProductoData): CrearProductoRequestDto {
    return {
      nombre: data.nombre,
      sku: data.sku,
      descripcion: data.descripcion,
      precio: data.precio,
      categoriaId: data.categoriaId,
      imagenNombre: data.imagenNombre,
    };
  }

  public static toUpdateRequest(data: EditarProductoData): ActualizarProductoRequestDto {
    return {
      id: data.id,
      nombre: data.nombre,
      sku: data.sku,
      descripcion: data.descripcion,
      precio: data.precio,
      categoriaId: data.categoriaId,
      imagenNombre: data.imagenNombre,
      activo: data.activo,
    };
  }
}
